libsnark/gadgetlib1/gadgets/hashes/anemoi/anemoi_components.tcc

```diff
@@ -0,0 +1,145 @@
+/** @file
+ *****************************************************************************
+
+ Implementation of interfaces for top-level Anemoi hash function gadgets.
+
+ See anemoi_gadget.hpp .
+
+ *****************************************************************************
+ * @author     This file is part of libsnark, developed by SCIPR Lab
+ *             and contributors (see AUTHORS).
+ * @copyright  MIT license (see LICENSE file)
+ *****************************************************************************/
+
+#ifndef LIBSNARK_GADGETLIB1_GADGETS_HASHES_ANEMOI_COMPONENTS_TCC_
+#define LIBSNARK_GADGETLIB1_GADGETS_HASHES_ANEMOI_COMPONENTS_TCC_
+
+namespace libsnark
+{
+
+// A R1CS constraint is a formal expression of the form
+//
+//                < A , X > * < B , X > = < C , X > ,
+//
+// where X = (x_0,x_1,...,x_m) is a vector of formal variables and
+// A,B,C each consist of 1+m elements in <FieldT> and <A, X> = \sum_i
+// (a_i x_i) is the dot product between vectors A and X. Equivalently,
+// the vectors A,B,C are linear combinations of X.
+//
+// A R1CS constraint is used to construct a R1CS constraint system.
+//
+// See also class \ef r1cs_constraint
+
+// R1CS constraints for the operation y = alpha x^2 + beta with x =
+// input, y = output. The latter is represented with one
+// multiplication as
+//
+// (alpha x) * x = y-beta
+//
+// for the variables vector X = (x0=1, x1=x, x2=y). This computation
+// is represented with 1 R1CS constraint:
+//
+// < A , X > * < B , X > = < C , X >
+//
+// where A =(0, alpha, 0), B=(0, 1, 0) and C =(-beta, 0, 1)
+template<typename FieldT>
+void anemoi_power_two_gadget<FieldT>::generate_r1cs_constraints()
+{
+    // X variables (input, output, intermediate) x_0=1, x_1, x_2
+    std::vector<pb_variable<FieldT>> X{ONE, input, output};
+    // A constants a_0, a_1, a_2
+    std::vector<FieldT> A{0, alpha, 0};
+    // B constants b_0, b_1, b_2
+    std::vector<FieldT> B{0, FieldT(1), 0};
+    // C constants c_0, c_1, c_2
+    std::vector<FieldT> C{-beta, 0, FieldT(1)};
+    // < A , X >
+    std::vector<linear_term<FieldT>> A_lc_terms{
+        {X[0], A[0]}, {X[1], A[1]}, {X[2], A[2]}};
+    linear_combination<FieldT> A_lc(A_lc_terms);
+    // < B , X >
+    std::vector<linear_term<FieldT>> B_lc_terms{
+        {X[0], B[0]}, {X[1], B[1]}, {X[2], B[2]}};
+    linear_combination<FieldT> B_lc(B_lc_terms);
+    // < C , X >
+    std::vector<linear_term<FieldT>> C_lc_terms{
+        {X[0], C[0]}, {X[1], C[1]}, {X[2], C[2]}};
+    linear_combination<FieldT> C_lc(C_lc_terms);
+    // < A , X > * < B , X > = < C , X >
+    this->pb.add_r1cs_constraint(
+        r1cs_constraint<FieldT>(A_lc, B_lc, C_lc),
+        FMT(this->annotation_prefix, " A*B=C"));
```

Comment:
I think this can be clarified a lot by removing the 0's and using the * and + operators on `linear_combination` to make the code look more like the equations:

```cpp
    // Constraint has the form:                   
    //   \alpha * input^2 + \beta = output        
    // which can be written as                    
    //   (\alpha * input) * input = output - \beta
    this->pb.add_r1cs_constraint(                 
        {input * alpha, input, output - beta},    
        FMT(this->annotation_prefix, " A*B=C"));  
```

Reply:
The annotation for the condition should be something meaningful.  "A*B=C" technically describes all r1cs conditions.

Reply:
Simplification applied -- thanks. Annotation fixed.